Save and load finitely presented groups coming from libgap groups

enriqueartal opened the pull request:
<!-- Describe your changes here in detail -->
At this point it is not possible to load a saved finitely presented group that comes from a libgap group, see an example in https://github.com/sagemath/sage/issues/37061

One possible cause is the use of a general `__reduce__` method for free groups. At least, adding such a method allows to load free groups or finitely presented groups obtained from a libgap group using `wrapFreeGroup` or `wrapFpGroup`. It fixes #37061 and it would simplify some code in https://github.com/sagemath/sage/pull/36768

With these changes, free and finitely presented groups, included libgap groups, can be pickled.

### :memo: Checklist



- [X] The title is concise, informative, and self-explanatory.
- [X] The description explains in detail what this PR is about.
- [X] I have linked a relevant issue or discussion.
- [X] I have created tests covering the changes.
- [X] I have updated the documentation accordingly.

enriqueartal committed:
wrap_FreeGroup

enriqueartal committed:
wrap_FpGroup

enriqueartal committed:
restore

enriqueartal committed:
new __reduce__ for free groups

enriqueartal committed:
new reduce

enriqueartal committed:
doctesting __reduce__

enriqueartal committed:
undoing strange changes

enriqueartal committed:
lint

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
eliminate wrap_FreeGroup

enriqueartal committed:
give up special __reduce__

enriqueartal committed:
sage method for gap f.p. groups

enriqueartal committed:
less wrap_FpGroup

enriqueartal committed:
no more wrap_FpGroup

enriqueartal committed:
trying...

enriqueartal committed:
changing sage method for gap f.p. groups and hom

enriqueartal committed:
Update free_group.py

typos and unnecessary wrap_FreeGroup

enriqueartal committed:
Update finitely_presented.py

enriqueartal committed:
Update free_group.py

Lint

enriqueartal committed:
Added gap groups for free and f.p groups, added __reduce__ methods
for these groups and also for artin, cubicbraid and raag

enriqueartal committed:
lint

enriqueartal committed:
merge

enriqueartal committed:
if the sage and gap generators are permuted, hom are correct

enriqueartal committed:
restore

enriqueartal committed:
change order of generators in Alternating groups to match the order in gap

enriqueartal committed:
private

enriqueartal committed:
.

enriqueartal committed:
.

enriqueartal committed:
CachedRepresentation

enriqueartal committed:
hash

enriqueartal committed:
doctests

enriqueartal committed:
change test in element.pyx

enriqueartal committed:
change test in element.pyx

enriqueartal committed:
add blank line

enriqueartal committed:
first attempt for richcmp

enriqueartal committed:
Merge branch 'homgroups' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into homgroups

tscrim reviewed (changes requested):
We are getting there. I don't really like free groups being a cached representation, but we are stuck with GAP's nonuniqueness choice and for FP groups. There still is a code smell here, but it is basically coming from GAP now. So we would need to change GAP to have better equality/uniqueness behaviors with FP and free groups to get rid of it.

Now some additional comments, but I think this will bring us to a positive review.

The free groups should also take in the libgap group, otherwise there is no point in downgrading them from a `UniqueRepresentation`. Subsequently, the libgap group should be passed in the GAP FPGroup `sage()` conversion code, and an analogous `sage()` conversion should be added for free groups. Although if there are actually no problems with having a unique free group within Sage, then make free groups back into a `UniqueRepresentation` (which I think is better).

Remove all changes in `braid.py`, `cubic_braid.py`, `permgroup_named.py`, and `raag.py` modulo one possible change: add `UniqueRepresentation` as the first class in the MRO. Your changes otherwise are not needed and just making the code more complicated. (Adding `UniqueRepresentation` puts them back with the uniqueness they deserve, and first in the MRO means the `UR.__reduce__` method will be used for the pickling.)

Your `_richcmp_` should be `__richcmp__` and you need a class-level decorator `@richcmp_method`. You should also properly test your comparisons that are not based on the cache: compare things with different libgap groups.

You should very clearly document in the class-level doc that any subclass of FPGroup will need to implement its own `__reduce__` method. I don't really like putting implementation details in public doc, but it is too important to not include it.

enriqueartal committed:
Merge branch 'groups' into homgroups

enriqueartal committed:
sage method for free groups and first attempt to richcmp_method

enriqueartal committed:
change reduce

enriqueartal committed:
forget richcmp

enriqueartal committed:
traling spaces

enriqueartal committed:
delete commented code

enriqueartal committed:
warning in doc

enriqueartal committed:
delete more commented code

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
merge develop

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
merge

enriqueartal committed:
-Change in unique representation to get a better mro
-Changes in free_group and finitely_presented to improve pickle
-Revert changes in artin, to do in raag and cubic_braid
-Changes in tests caused by the previous ones, still problems with cartan_type

enriqueartal committed:
take out reduce and add UniqueRepresentation

enriqueartal committed:
lint

enriqueartal committed:
more doctests

enriqueartal committed:
lint

enriqueartal committed:
minor change in one test

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
merge

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
merge 10.4beta4

enriqueartal committed:
deleting problematic backward tests

enriqueartal committed:
merge

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

tscrim commented on src/sage/groups/finitely_presented.py:
```diff
@@ -786,6 +745,25 @@ def __init__(self, free_group, relations, category=None):
+            ...
+            ValueError: variable names have not yet been set using self._assign_names(...)
+
+        Check that pickling works
```
```suggestion
        Check that pickling works::
```

tscrim commented on src/sage/schemes/curves/zariski_vankampen.py:
```diff
@@ -1543,8 +1542,8 @@ def braid2rels(L):
-        gb = wrap_FpGroup(P.FpGroupPresentation())
-        U = [rel.Tietze() for rel in gb.relations()]
+        gb = P.FpGroupPresentation().sage()
+        U = [_.Tietze() for _ in gb.relations()]
```
```suggestion
        U = [rel.Tietze() for rel in gb.relations()]
```
The `_` is generally meant for cases when the iterating object is not used in the iterator itself, e.g., `[1 for _ in range(5)]`.

tscrim commented on src/sage/groups/finitely_presented.py:
```diff
@@ -794,11 +772,45 @@ def __init__(self, free_group, relations, category=None):
+            sage: G2 = FreeGroup(2) / [(1, 2, 2, 1, 2, 1)]
+            sage: G1 == G2
+            True
+        """
```
```suggestion
            True
            sage: G1 is G2
            False
        """
```

enriqueartal replied:
Actually I got `True` for `G1 is G2`

tscrim replied:
Then currently this test is not good. It should be comparing equal-but-not-identical groups. (Actually, we might want to have a shortcut for identical groups, but that might be a premature optimization.)

enriqueartal replied:
You are right, the groups `G1` and `G2` are defined exactly in the same way. I tried several ways to produce `True` for `==` and `False` for `is` with other definitions but I failed.

tscrim replied:
How about passing the gap group?

enriqueartal replied:
It worked defining `G2 = libgap(G1).sage()`

tscrim commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -716,6 +716,7 @@ def __init__(self, domain=None):
             sage: h = A.hom(codomain=A, im_gens=A.gens())
         """
         PermutationGroup_symalt.__init__(self, gap_group='AlternatingGroup(%s)' % len(domain), domain=domain)
+        self._gens = tuple(self(g) for g in self.gap().GeneratorsOfGroup())
```
Is this still needed?

enriqueartal replied:
Apparently not.

tscrim commented on src/sage/groups/raag.py:
```diff
@@ -38,6 +38,7 @@
 from sage.categories.groups import Groups
 from sage.categories.algebras_with_basis import AlgebrasWithBasis
 from sage.algebras.clifford_algebra_element import CohomologyRAAGElement
+from sage.structure.unique_representation import CachedRepresentation
```
```suggestion
```
This is no longer needed.

tscrim reviewed (changes requested):
Sorry, I haven't had much time to get to this until now.

I am still really bothered by the change breaking the Cartan type unpickling. I really do not understand why that doesn't work. Hence, I am very worried we could unknowingly be breaking something that isn't tested otherwise. It's probably okay since it is for *very* old pickles. I just wish I knew what was breaking there...

Maybe for better transparency, we should split the `UniqueRepresentation` change off as a separate PR. Would you be okay doing this? Sorry for the extra work.

enriqueartal committed:
small changes

enriqueartal committed:
tests in richcmp

enriqueartal committed:
merge minor conflicts

enriqueartal committed:
typo

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
change tests due to unique representation change

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

enriqueartal committed:
Merge branch 'sagemath:develop' into groups

tscrim reviewed (approved):
Sorry for the very long wait in getting to this and finishing my review. I have no additional comments. Let's get this in. Again, sorry for not getting this done sooner.
